File: src/Catty/ViewController/CatrobatTableViewController/SettingsTableViewController.m
/**
 *  Copyright (C) 2010-2020 The Catrobat Team
 *  (http://developer.catrobat.org/credits)
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Affero General Public License as
 *  published by the Free Software Foundation, either version 3 of the
 *  License, or (at your option) any later version.
 *
 *  An additional term exception under section 7 of the GNU Affero
 *  General Public License, version 3, is available at
 *  (http://developer.catrobat.org/license_additional_term)
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *  GNU Affero General Public License for more details.
 *
 *  You should have received a copy of the GNU Affero General Public License
 *  along with this program.  If not, see http://www.gnu.org/licenses/.
 */

#import "SettingsTableViewController.h"
#import "TermsOfUseOptionTableViewController.h"
#import "AboutPoketCodeOptionTableViewController.h"
#import "LanguageTranslationDefines.h"
#import "KeychainUserDefaultsDefines.h"
#import "CatrobatTableViewController.h"
#import "Pocket_Code-Swift.h"

@implementation SettingsTableViewController

- (void)setup {
    
    self.title = kLocalizedSettings;
    self.view.backgroundColor = UIColor.background;
    self.view.tintColor = UIColor.globalTint;
    
    [self addSection:[BOTableViewSection sectionWithHeaderTitle:@"" handler:^(BOTableViewSection *section) {
        
        if ([Util isPhiroActivated]) {
            [section addCell:[BOSwitchTableViewCell cellWithTitle:kLocalizedPhiroBricks key:kUsePhiroBricks handler:^(BOSwitchTableViewCell *cell) {
                cell.backgroundColor = UIColor.background;
                cell.mainColor = UIColor.globalTint;
                cell.toggleSwitch.tintColor = UIColor.globalTint;
                [cell.toggleSwitch setOnTintColor:UIColor.globalTint];
            }]];
        }
        
        if ([Util isArduinoActivated]) {
            [section addCell:[BOSwitchTableViewCell cellWithTitle: kLocalizedArduinoBricks key:kUseArduinoBricks handler:^(BOSwitchTableViewCell *cell) {
                cell.backgroundColor = UIColor.background;
                cell.mainColor = UIColor.globalTint;
                cell.toggleSwitch.tintColor = UIColor.globalTint;
                [cell.toggleSwitch setOnTintColor:UIColor.globalTint];
                cell.onFooterTitle = kLocalizedArduinoBricksDescription;
                cell.offFooterTitle = kLocalizedArduinoBricksDescription;
            }]];
        }
        
    }]];
    __unsafe_unretained typeof(self) weakSelf = self;
    BluetoothService *service = [BluetoothService sharedInstance];
    
    if (([Util isPhiroActivated] || [Util isArduinoActivated]) ) {
        [self addSection:[BOTableViewSection sectionWithHeaderTitle:@"" handler:^(BOTableViewSection *section) {
            if((service.phiro != nil || service.arduino != nil)){
                [section addCell:[BOButtonTableViewCell cellWithTitle:kLocalizedDisconnectAllDevices key:nil handler:^(BOButtonTableViewCell *cell) {
                    cell.backgroundColor = UIColor.background;
                    cell.mainColor = UIColor.globalTint;
                    cell.actionBlock = ^{
                        [weakSelf disconnect];
                    };
                }]];
            }
            
            NSArray *tempArray = [[NSUserDefaults standardUserDefaults] arrayForKey:@"KnownBluetoothDevices"];
            if(tempArray.count) {
                [section addCell:[BOButtonTableViewCell cellWithTitle:kLocalizedRemoveKnownDevices key:nil handler:^(BOButtonTableViewCell *cell) {
                    cell.backgroundColor = UIColor.background;
                    cell.mainColor = UIColor.globalTint;
                    cell.actionBlock = ^{
                        [weakSelf removeKnownDevices];
                    };
                }]];
                
            }
            
        }]];
        
    }
    
    if ([[[NSUserDefaults standardUserDefaults] valueForKey:kUserIsLoggedIn] boolValue])
    {
        [self addSection:[BOTableViewSection sectionWithHeaderTitle:@"" handler:^(BOTableViewSection *section) {
            [section addCell:[BOButtonTableViewCell cellWithTitle:kLocalizedLogout key:nil handler:^(BOButtonTableViewCell *cell) {
                cell.backgroundColor = UIColor.background;
                cell.mainColor = UIColor.variableBrickRed;
                cell.actionBlock = ^{
                    [weakSelf logoutUser];
                    [self.navigationController popViewControllerAnimated:YES];
                };
            }]];
        }]];
    }
    
    [self addSection:[BOTableViewSection sectionWithHeaderTitle:@"" handler:^(BOTableViewSection *section) {
        [section addCell:[BOChoiceTableViewCell cellWithTitle:kLocalizedAboutPocketCode key:@"choice_2" handler:^(BOChoiceTableViewCell *cell) {
            cell.destinationViewController = [AboutPoketCodeOptionTableViewController new];
            cell.backgroundColor = UIColor.background;
            cell.mainColor = UIColor.globalTint;
        }]];
        [section addCell:[BOChoiceTableViewCell cellWithTitle:kLocalizedTermsOfUse key:@"choice_2" handler:^(BOChoiceTableViewCell *cell) {
            cell.destinationViewController = [TermsOfUseOptionTableViewController new];
            cell.backgroundColor = UIColor.background;
            cell.mainColor = UIColor.globalTint;
        }]];
    }]];
    
    [self addSection:[BOTableViewSection sectionWithHeaderTitle:@"" handler:^(BOTableViewSection *section) {
        
        [section addCell:[BOButtonTableViewCell cellWithTitle:kLocalizedPrivacySettings key:nil handler:^(BOButtonTableViewCell *cell) {
            cell.backgroundColor = UIColor.background;
            cell.mainColor = UIColor.globalTint;
            cell.actionBlock = ^{
                [weakSelf openPrivacySettings];
            };
        }]];
        [section addCell:[BOButtonTableViewCell cellWithTitle:kLocalizedRateUs key:nil handler:^(BOButtonTableViewCell *cell) {
            cell.backgroundColor = UIColor.background;
            cell.mainColor = UIColor.globalTint;
            cell.actionBlock = ^{
                [weakSelf openRateUsURL];
            };
        }]];
        
        NSString *version = [[NSString alloc] initWithFormat:@"%@%@ (%@)", kLocalizedVersionLabel,
        [[[NSBundle mainBundle] infoDictionary] objectForKey:@"CFBundleShortVersionString"], [Util appBuildVersion]];
        
        #if DEBUG == 1
            version = [NSString stringWithFormat: @"%@(%@)", version, kLocalizedDebugMode];
        #endif
        
        section.footerTitle = version;
    }]];
}

-(void)viewWillDisappear:(BOOL)animated{
    [super viewWillDisappear:animated];
    [[NSUserDefaults standardUserDefaults] synchronize];
}

- (void)presentAlertControllerWithTitle:(NSString *)title message:(NSString *)message {
    [Util alertWithTitle:title andText:message];
}

- (void)showTappedButtonAlert {
    // open url
    [self presentAlertControllerWithTitle:@"Button tapped!" message:nil];
}

- (void)openRateUsURL
{
    [[UIApplication sharedApplication] openURL:[NSURL URLWithString:NetworkDefines.appStoreUrl]];
}

- (void)openPrivacySettings
{
    [[UIApplication sharedApplication] openURL:[NSURL URLWithString:UIApplicationOpenSettingsURLString]];
}

- (void)disconnect
{
    [[BluetoothService sharedInstance] disconnect];
    [Util alertWithText:kLocalizedDisconnectBluetoothDevices];
}

- (void)removeKnownDevices
{
    [[BluetoothService sharedInstance] removeKnownDevices];
    [Util alertWithText:kLocalizedRemovedKnownBluetoothDevices];
}

- (void)logoutUser
{
    [[NSUserDefaults standardUserDefaults] setValue:false forKey:kUserIsLoggedIn];
    [[NSUserDefaults standardUserDefaults] setValue:@"" forKey:kUserLoginToken];
    [[NSUserDefaults standardUserDefaults] setValue:@"" forKey:kcUsername];
}

@end

File: src/Catty/ViewController/CatrobatTableViewController/SettingsTableViewController.swift
/**
 *  Copyright (C) 2010-2023 The Catrobat Team
 *  (http://developer.catrobat.org/credits)
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Affero General Public License as
 *  published by the Free Software Foundation, either version 3 of the
 *  License, or (at your option) any later version.
 *
 *  An additional term exception under section 7 of the GNU Affero
 *  General Public License, version 3, is available at
 *  (http://developer.catrobat.org/license_additional_term)
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *  GNU Affero General Public License for more details.
 *
 *  You should have received a copy of the GNU Affero General Public License
 *  along with this program.  If not, see http://www.gnu.org/licenses/.
 */

import Foundation

@objc class SettingsTableViewController: FormTableViewController {

    private let sectionHeaders: [String?] = [
        kLocalizedExtensions, kLocalizedPrivacy, kLocalizedAbout, kLocalizedMore, nil
    ]

    private let sectionFooters: [String?] = [
        nil, nil, nil, nil, nil
    ]

    private var currentSectionHeaders: [String?] = []
    private var currentSectionFooters: [String?] = []

    private var featureItems: [FormItem] = []
    private var bluetoothItems: [FormItem] = []
    private var webAccessItems: [FormItem] = []
    private let authenticator = StoreAuthenticator()

    // MARK: - Lifecycle

    override func viewDidLoad() {
        super.viewDidLoad()

        setupViews()
        setupFormItems()
    }

    // MARK: - UI

    private func setupViews() {
        self.tableView.backgroundColor = .background
        self.navigationItem.title = kLocalizedSettings
    }

    // MARK: - TableView

    override func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        currentSectionHeaders[section]
    }

    override func tableView(_ tableView: UITableView, titleForFooterInSection section: Int) -> String? {
        currentSectionFooters[section]
    }

    // MARK: - Functions

    private func setupFormItems() {
        self.currentSectionHeaders = self.sectionHeaders
        self.currentSectionFooters = self.sectionFooters

        setupFeatureItems()
        setupBluetoothItems()
        setupWebAccessItems()

        formItems = [ self.featureItems,
            [
                FormCrashReportsSwitchItem(switchAction: { isEnabled in
                    self.changeFirebaseCrashReportSettings(isEnabled: isEnabled)
                }),
                FormItem(title: kLocalizedPrivacySettings, titleColor: .globalTint, action: {
                    self.openPrivacySettings()
                })
            ],
            [
                FormItem(title: kLocalizedAboutUs, accessoryType: .disclosureIndicator, action: {
                    self.showAboutUs()
                })
            ],
            [
                FormVersionItem(),
                FormItem(title: kLocalizedTermsOfUse, accessoryType: .disclosureIndicator, action: {
                    self.showTermsOfUseAndService()
                }),
                FormItem(title: kLocalizedRateUs, titleColor: .globalTint, action: {
                    self.openRateUsURL()
                })
            ]
        ]

        if UserDefaults.standard.bool(forKey: kUseWebRequestBrick) {
            formItems.insert(self.webAccessItems, at: 1)
            currentSectionHeaders.insert(kLocalizedWebAccess, at: 1)
            currentSectionFooters.insert(nil, at: 1)
        }

        if UserDefaults.standard.bool(forKey: kUseArduinoBricks), !bluetoothItems.isEmpty {
            formItems.insert(self.bluetoothItems, at: 1)
            currentSectionHeaders.insert(kLocalizedArduinoBricks, at: 1)
            currentSectionFooters.insert(nil, at: 1)
        }

        if StoreAuthenticator.isLoggedIn() {
            formItems.append([
                FormItem(title: kLocalizedLogout, titleColor: .red, action: {
                    self.logout()
                }),
                FormItem(title: kLocalizedDeleteAccount, titleColor: .red, action: {
                    self.deleteAccount()
                })
            ])
        }
    }

    private func setupFeatureItems() {
        self.featureItems = []

        if Util.isPhiroActivated() {
            self.featureItems.append(FormPhiroSwitchItem())
        }

        if Util.isArduinoActivated() {
            self.featureItems.append(FormArduinoSwitchItem(switchAction: { _ in
                self.setupFormItems()
            }))
        }

        if Util.isEmbroideryActivated() {
            self.featureItems.append(FormEmbroiderySwitchItem())
        }
    }

    private func setupBluetoothItems() {
        self.bluetoothItems = []

        if BluetoothService.sharedInstance().phiro != nil || BluetoothService.sharedInstance().arduino != nil {
            self.bluetoothItems.append(
                FormItem(title: kLocalizedDisconnectAllDevices, action: {
                    self.disconnectAllBluetoothDevices()
                })
            )
        }

        if let knownDevices = UserDefaults.standard.array(forKey: kKnownBluetoothDevices), !knownDevices.isEmpty {
            self.bluetoothItems.append(
                FormItem(title: kLocalizedRemoveKnownDevices, action: {
                    self.removeKnownBluetoothDevices()
                })
            )
        }
    }

    private func setupWebAccessItems() {
        self.webAccessItems = [
            FormItem(title: kLocalizedTrustedDomains, accessoryType: .disclosureIndicator, action: {
                self.showTrustedDomains()
            })
        ]
    }

    func changeFirebaseCrashReportSettings(isEnabled: Bool) {
         NotificationCenter.default.post(
            name: NSNotification.Name(rawValue: NotificationName.settingsCrashReportingChanged),
            object: NSNumber(value: isEnabled))
     }

    private func disconnectAllBluetoothDevices() {
        BluetoothService.sharedInstance().disconnect()
        Util.alert(text: kLocalizedDisconnectBluetoothDevices)
    }

    private func removeKnownBluetoothDevices() {
        BluetoothService.sharedInstance().removeKnownDevices()
        Util.alert(text: kLocalizedRemovedKnownBluetoothDevices)
    }

    private func openRateUsURL() {
        if let url = URL(string: NetworkDefines.appStoreUrl) {
            UIApplication.shared.open(url, options: [:], completionHandler: nil)
        }
    }

    private func openPrivacySettings() {
        if let url = URL(string: UIApplication.openSettingsURLString) {
            UIApplication.shared.open(url, options: [:], completionHandler: nil)
        }
    }

    private func showTrustedDomains() {
        self.navigationController?.pushViewController(TrustedDomainTableViewController(), animated: true)
    }

    private func showTermsOfUseAndService() {
        self.navigationController?.pushViewController(TermsOfUseTableViewController(), animated: true)
    }

    private func showAboutUs() {
        self.navigationController?.pushViewController(AboutPocketCodeTableViewController(), animated: true)
    }

    private func logout() {
        StoreAuthenticator.logout()
        self.navigationController?.popViewController(animated: true)
    }

    private func deleteAccount() {
        let alertController = UIAlertController(title: kLocalizedDeleteAccount, message: kLocalizedDeleteAccountConfirm, preferredStyle: .alert)
        let cancelAction = UIAlertAction(title: kLocalizedCancel, style: .cancel, handler: nil)
        let deleteAction = UIAlertAction(title: kLocalizedDelete, style: .destructive) { _ in
            self.authenticator.deleteUser { error in
                DispatchQueue.main.async(execute: {
                    switch error {
                    case .none:
                        self.navigationController?.popViewController(animated: true)
                        Util.alert(text: kLocalizedDeleteAccountSuccessful)
                    case .authentication:
                        Util.alert(text: kLocalizedAuthenticationFailed)
                    case .network, .timeout:
                        Util.defaultAlertForNetworkError()
                    default:
                        Util.alert(text: kLocalizedUnexpectedErrorMessage)
                    }
                })
            }
        }
        alertController.addAction(cancelAction)
        alertController.addAction(deleteAction)
        present(alertController, animated: true, completion: nil)
    }
}
